web/packages/teleterm/src/mainProcess/agentDownloader/fileDownloader.ts

```diff
@@ -0,0 +1,124 @@
+/**
+ * Copyright 2023 Gravitational, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import path from 'node:path';
+
+import { BrowserWindow, DownloadItem } from 'electron';
+
+import Logger from 'teleterm/logger';
+
+export interface IFileDownloader {
+  run(url: string, downloadDirectory: string): Promise<void>;
+}
+
+export class FileDownloader implements IFileDownloader {
+  private logger = new Logger('fileDownloader');
+  private onResolve: () => void;
+  private onReject: (error: Error) => void;
+
+  constructor(private window: BrowserWindow) {}
+
+  async run(url: string, downloadDirectory: string) {
+    this.logger.info(
+      `Starting download from ${url} (download directory: ${downloadDirectory}).`
+    );
+    const handler = this.createDownloadHandler(url, downloadDirectory);
+    this.window.webContents.session.on('will-download', handler);
+
+    try {
+      return await new Promise<void>((resolve, reject) => {
+        this.onResolve = resolve;
+        this.onReject = reject;
```

Comment:
Personally I'd call `this.createDownloadHandler(url, downloadDirectory)` from within the promise body so that I'm able to pass `resolve` and `reject` directly to it rather than depending on the state of a class instance.

But I'm fine with either way, it's just that the code is easier to follow if it doesn't depend on some state being defined in a completely different place.

Reply:
Yeah, it makes sense to remove unneeded state. When I started working on this I put the state in class and didn't think later of removing it.